### PR TITLE
Model il

### PR DESCRIPTION
**Pull request by jepidoptera**

### CODE REVIEW

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    "from model.amm.agents import Agent\n",
+    "from model.amm.trade_strategies import omnipool_arbitrage, invest_all, price_sensitive_trading\n",
+    "from model.amm.global_state import GlobalState, fluctuate_prices, historical_prices\n",
+    "\n",
+    "# same seed, same parameters = same simulation result\n",
+    "random.seed(42)\n",
+    "# price_list = processing.import_binance_prices(['BTC', 'ETH', 'DOT'], start_date='Jan 1 2023', days = 120)\n",
+    "\n",
+    "assets = {\n",
+    "    'HDX': {'usd price': 0.05, 'weight': 0.7999},\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:32]: HDX plays a special role in the Omnipool currently, with LRNA fees being directed to the HDX pool whenever imbalance = 0. I'd thus prefer HDX weight to be realistic (5% or so?) and for there to be some other generic token to make up the additional 65% of the pool.
Would this wind up being identical?

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    "\n",
+    "initial_state = GlobalState(\n",
+    "    pools={\n",
+    "        'omnipool': OmnipoolState(\n",
+    "            tokens={\n",
+    "                tkn: {'liquidity': liquidity[tkn], 'LRNA': lrna[tkn]} for tkn in assets\n",
+    "            },\n",
+    "            asset_fee=0,\n",
+    "            lrna_fee=0,\n",
+    "            # withdrawal_fee=False\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:54]: Does this mean that the withdrawal_fee is *on* in this analysis?

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    "            )\n",
+    "        }\n",
+    "\n",
+    "        events.append([])\n",
+    "        for price in price_move_range:\n",
+    "            price_list = [{'TKN': initial_state.external_market['TKN']}, {'TKN': price}]\n",
+    "            initial_state._evolve_function = historical_prices(price_list)\n",
+    "            events[w].append(run.run(initial_state, 1, silent=True)[-1])\n",
+    "\n",
+    "    impermanent_loss = [[\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:153]: I think it's better to say that e.g. if the price goes up 90%, the "impermanent loss" is 5% (or -5% depending on sign convention). This variable is dividing the cash_out value by the hold value, so "impermanent_loss" is 95%. Everything is correct but the variable name should probably not be "impermanent_loss" since it's actually the value *retained* for the LP during a price move.

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    "from model.amm.agents import Agent\n",
+    "from model.amm.trade_strategies import omnipool_arbitrage, invest_all, price_sensitive_trading\n",
+    "from model.amm.global_state import GlobalState, fluctuate_prices, historical_prices\n",
+    "\n",
+    "# same seed, same parameters = same simulation result\n",
+    "random.seed(42)\n",
+    "# price_list = processing.import_binance_prices(['BTC', 'ETH', 'DOT'], start_date='Jan 1 2023', days = 120)\n",
+    "\n",
+    "assets = {\n",
+    "    'HDX': {'usd price': 0.05, 'weight': 0.7999},\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb:32]: Same comment as other notebook

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    "        other_assets = ['HDX']\n",
+    "        for price in price_move_range:\n",
+    "            price_list = [\n",
+    "                {tkn: initial_state.external_market[tkn] for tkn in other_assets}, \n",
+    "                {tkn: initial_state.external_market[tkn] * price for tkn in other_assets}\n",
+    "            ]\n",
+    "            initial_state._evolve_function = historical_prices(price_list)\n",
+    "            events[w].append(run.run(initial_state, 1, silent=True)[-1])\n",
+    "\n",
+    "    impermanent_loss = [[\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb:155]: Same comment as other notebook

**Review by poliwop — CHANGES_REQUESTED** [review]

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb (hunk 2 above)
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:54]: There's no withdrawal fee, but that's because OmnipoolArchiveState doesn't include that attribute. That's probably something that should be addressed? It certainly makes a case like this unclear.

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb (hunk 3 above)
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:153]: That is a good point, I'll change it.

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb (hunk 4 above)
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb:32]: For this one, I agree, so I changed it. It didn't make a noticeable difference, but I think it's more correct in principle.

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb (hunk 2 above)
Comment (poliwop) [reply hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:54]: So is the withdrawal fee on or off in this analysis?

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb (hunk 1 above)
Comment (poliwop) [reply hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:32]: The HDX/USD price shouldn't change in this notebook, right? Is that what we see? I would expect there needs to be some trade involving HDX to make this happen, which would change its weight?

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    "    price_move_range = [price_min * step_factor ** i for i in range(steps)]\n",
+    "    events = []\n",
+    "    weight_factors = [0.0001, 0.01, 0.02, 0.035, 0.05]\n",
+    "\n",
+    "\n",
+    "    for w, weight in enumerate(weight_factors):\n",
+    "        assets = {\n",
+    "            'HDX': {'usd price': 0.05, 'weight': 0.05},\n",
+    "            'USD': {'usd price': 1, 'weight': 0.8 - weight},\n",
+    "            'TKN': {'usd price': 1, 'weight': weight}\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:131]: Shouldn't this have all 3 tokens, i.e. shouldn't this have TKN2 as well?

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb (hunk 1 above)
Comment (poliwop) [reply hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb]: So my question above stands... and I see that we have TKN and TKN2 in both notebooks? That's what I believe is appropriate.

**Review by poliwop — CHANGES_REQUESTED** [review]
I think something is off in ImpermanentLossAnalysis.ipynb between the initial instantiation of the Omnipool and the instantiation in the for look. TKN2 is initially in the pool, but isn't present in the for loop. It's not clear to me which is correct.

In ImpermanentLossAnalysis2.ipynb, I think we need HDX to move with the market too. This may mean that I was wrong to suggest that we *need* HDX and TKN2 to be split out separately, although I still think this is actually clearer. I would simply add HDX to the "other_assets" so that it moves with the market.

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb (hunk 6 above)
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis.ipynb:131]: You're right, this is making the weights wrong

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb [hunk 7]
```diff
@@ -159,7 +159,7 @@
     "\n",
```
Comment (poliwop) [inline hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb:159]: Shouldn't "other_assets" now be ['TKN2']? I think as it stands this notebook is changing the price of TKN, not TKN2?

**Review by poliwop — CHANGES_REQUESTED** [review]
I think one more change is needed in ImpermanentLossAnalysis2.ipynb

File: hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb (hunk 7 above)
Comment (jepidoptera) [reply hydradx/notebooks/Omnipool/vault/ImpermanentLossAnalysis2.ipynb:159]: You were right, fixed that.

**Review by poliwop — APPROVED** [review]
